### ISSUE TITLE
[Crash] Crashing calling ‘Change Setting’ from system menu

### ISSUE THREAD

**Issue by kissglay** [cyd01/KiTTY#113]
Click ‘Change Setting’ from system menu and an empty setting window appear, then crashing.
This is problem existing for quite a long time.
![2020-02-15_16 18 50 923__KiTTY](https://user-images.githubusercontent.com/529385/74584622-6952ff80-500f-11ea-8ebc-49a6bee21890.png)
![2020-02-15_16 19 01 746](https://user-images.githubusercontent.com/529385/74584624-740d9480-500f-11ea-86b9-8d73d5afbeed.png)

**Comment by Dubbeldrank** [comment]
I can confirm this, using Windows 10 build 1909.

**Comment by cyd01** [comment]
I need more informations on your session to try to reproduce the issue.

**Comment by kissglay** [comment]
I test with "Default setting" and the same problem.
I remembered this issue was introduced from  0.73, and I once reported this issue through your web site (i.e. "kitty.9bis.net").

**Comment by cyd01** [comment]
Did you try to remove your "Default Settings" registry entry (or file if
you use portable mode) ?

Le dim. 23 févr. 2020 à 05:24, Max33Verstappen <notifications@github.com> a
écrit :

**Comment by kissglay** [comment]
Yes. I remove that entry from the registry and start purely clean, and exactly the same issue.

BTW, at that time I also reported an issue relating kitty's window refreshing / redrawing issue (very slow refreshing from top left to bottom right). I dunno if they're related somehow.

**Comment by cyd01** [comment]
Disable background image feature: see this page:
http://www.9bis.net/kitty/#!pages/covidimus.md


Le dim. 23 févr. 2020 à 14:26, Max33Verstappen <notifications@github.com> a
écrit :

**Comment by kissglay** [comment]
I never used this feature.

I found this issue changed slightly in version 0.73.2.4. When I hit "Change Setting...", the option dialog window at least showed, but after a couple of seconds (random), it crashes.

**Comment by cyd01** [comment]
Without reproducing the issue it will be hard to fix it!
Can you send me a memory dump: http://www.9bis.net/kitty/#!pages/Savedump.md


Le dim. 23 févr. 2020 à 16:03, Max33Verstappen <notifications@github.com> a
écrit :

**Comment by kissglay** [comment]
I attached the dump file as per the instruction.
This is from windows event viewer:
Faulting application name: kitty-debug.exe, version: 0.73.2.3, time stamp: 0x5e5279fc
Faulting module name: kitty-debug.exe, version: 0.73.2.3, time stamp: 0x5e5279fc
Exception code: 0xc0000005
Fault offset: 0x0006f9cb
Faulting process id: 0x5018
Faulting application start time: 0x01d5ea5e5d8427fe
Faulting application path: C:\tools\Putty\kitty-debug.exe
Faulting module path: C:\tools\Putty\kitty-debug.exe
Report Id: 8469f986-bb1f-4273-9602-79dde92bc696
Faulting package full name: 
Faulting package-relative application ID: 
[kitty.dmp.zip](https://github.com/cyd01/KiTTY/files/4241508/kitty.dmp.zip)

**Comment by cyd01** [comment]
I'm sorry but I was not able to reproduce.
I've imported all your registry settings, and I've used your kitty.ini
configuration file.
But unfortunately I was not able to reproduce the crash (going to "change
setting" menu several times).
Maybe KiTTY enters in conflict with some of the other softwares that run on
your system.
I suggest you to try to disable them one by one.



Le dim. 23 févr. 2020 à 16:39, Max33Verstappen <notifications@github.com> a
écrit :

**Comment by kissglay** [comment]
Strange...
I don't have this issue on other 2 PCs, which are on the almost same software environment as the one with this issue, but those are INTEL CPUs.
The one with the issue is an AMD.
Donno if it's CPU specific.

But I don't have this issue with the original putty.
And I find this issue only occurs when the session is active. When the session is inactive (disconnected or whatever) , I can open the setting from system menu.

**Comment by Dubbeldrank** [comment]
I'm also an AMD user for your interest.

**Comment by piksel** [comment]
Same problem here. Also AMD CPU if that is relevant (Ryzen 7 3800X).

**Comment by n0rc** [comment]
I have the same problem, tested with 0.73.2.2 and also with 0.73.2.9.

**Comment by thebertster** [comment]
I have the same problem with 0.73.2.10 on my PC with AMD Ryzen 5 CPU (and nVidia GPU). I do not have the issue on my laptop which has an Intel CPU.

The crash occurs when the KiTTY window is directly below the mouse pointer. In other words, the crash does not occur if the Settings window is opened such that the mouse pointer is either outside of the KiTTY window or the mouse pointer is within the Settings window when it opens (which may explain why some people are saying the crash is intermittent) - but if the mouse pointer is then moved to be directly over the KiTTY window, the crash occurs.

**Comment by piksel** [comment]
Confirmed the behaviour that @thebertster described. It seems like the `WM_MOUSEMOVE` event on the main window causes an `ACCESS_VIOLATION` exception to occur when the settings window is open.

**Comment by thebertster** [comment]
I tested with old versions and the problem seems to have been introduced in 71.0.1.

**Comment by cyd01** [comment]
Can someone run this [debug build](https://www.9bis.net/kitty/files/kitty-debug.exe) ?
It should create a **kitty.log** file in the same directory where kitty.exe is.
Then contact me through the [contact page](https://www.9bis.net/kitty/#!pages/contact.md) of the official web site. With the log file, I'll try to determine where the crash is located.

**Comment by thebertster** [comment]
Done.

**Comment by kissglay** [comment]
Unfortunately, for me, the problem remains in v0.73.2.11

**Comment by sarbian** [comment]
I have the same problem with 0.73.2.11. Do you need more logs with the debug build ?

**Comment by thebertster** [comment]
We went back and forth multiple times with debug builds trying to get to the bottom of this. The problem is that the crash doesn't seem to be happening in the KiTTY code itself.

So far, every report of this has involved AMD CPUs - is that the case also for you @kissglay amd @sarbian? If so, my instinct is that the issue is with the actual compiler or mingw libraries. There have been a number of AMD-specific issues with it in the past - though we tried the workarounds for these to no avail.

One other thing to note is that I found the 64-bit build does not crash - might be worth you trying that - I believe it's available now on FossHub?

**Comment by bompus** [comment]
The last 64-bit build that I see on FossHub is 0.73.2.4

**Comment by sarbian** [comment]
Yes, AMD CPU here.
The 64-build on FossHub seems to work. 

I see that github action is configured to build the 64 build but it fails with an error about a wrong cpu type. If it was fixed it would provide a (hopefully) working  build for AMD users.

**Comment by nanaya** [comment]
Another AMD cpu having similar crash. Version 0.73.2.10, running on Ryzen 5 3600, Windows 10 (1909).

**Comment by cyd01** [comment]
64 bits version will be back soon. But it was reported to be buggy.

**Comment by cyd01** [comment]
> 
> 
> The last 64-bit build that I see on FossHub is 0.73.2.4. It also seems to be the full version, while I primarily use the portable.



> 
> 
> The last 64-bit build that I see on FossHub is 0.73.2.4. It also seems to be the full version, while I primarily use the portable.

Portable mode does not really exists. In fact there only one KiTTY version, that you can configure in **classic mode** or **portable mode**. See [this page](https://www.9bis.net/kitty/#!pages/Portability.md).

**Comment by cyd01** [comment]
Can someone confirm that the crash occurs with all KiTTY versions:
* from [fosshub](https://www.fosshub.com/KiTTY.html)
    * classic mode
    * portable mode
    * uncompressed mode
* from [Github actions](https://github.com/cyd01/KiTTY/actions/runs/97104143)
    * 32 bits
    * 64 bits

Thanks for your help.

**Comment by nanaya** [comment]
My testing steps:
1. open `kitty.exe` (or `_nocompress`, etc)
2. enter ip address
3. select ssh
4. open
5. right click title bar
6. select "Change settings"

Result:
Crash on everything except for `kitty64.exe` from Github actions 64 bits.

**Comment by cyd01** [comment]
Thanks for your help

**Comment by ashemedai** [comment]
Ryzen 9 CPU here, had the problem (#176 reporter). Upgraded to 0.73.3.12 today which showed the crash. With 0.73.1.1 I did *not* have this problem. It also somehow threw away all my stored sessions after launching the replaced binary, not sure if this might be relevant.

**Comment by thebertster** [comment]
I previously tested all the different versions and see the same - all but the 64-bit build crash in the same way.

**Comment by bompus** [comment]
Ryzen 3950X, I could not get the 64bit version from Github actions to crash on me when changing settings.

**Comment by kissglay** [comment]
Tested on Threadripper 1950x, 64bit build doesn't suffer crash issue when opening "changing setting" on the fly.

**Comment by ashemedai** [comment]
As others stated, the 64-bit GitHub actions build also works for me.

**Comment by cyd01** [comment]
@ashemedai you confirm the issue appears between 0.73.1.1 and 0.73.1.2 ?

**Comment by ashemedai** [comment]
> @ashemedai you confirm the issue appears between 0.73.1.1 and 0.73.1.2 ?

I am sorry, I have to take that back. I was sure that I had used that 0.73.1.1 version that I had prior to the upgrade with Change Settings, but I just downloaded that version from Fosshub and it fails for me as well.

Since I am busy with testing now:

| version | works |
| --- | --- |
| 0.73.3.12 | ⛔ | 
| 0.73.1.1 | ⛔ |
| 0.72.0.1 | ⛔ |
| 0.71.0.7 | ⛔ |
| 0.71.0.5 | ⛔ |
| 0.71.0.4 | ✅ |
| 0.71.0.1 | ✅ |
| 0.70.0.10 | ✅ |

So it looks like something in the build from 0.71.0.4 to 0.71.0.5 causes the exception violations.
Hope this helps!

**Comment by thebertster** [comment]
@ashemedai - can you double-check 0.71.0.4 and 0.71.0.1 again and make sure that your mouse is moving over the main KiTTY window and not staying within the settings window? I have tested all the 0.71.x versions multiple times and they all consistently crash for me when the mouse moves over the main KiTTY window when Settings is open. The last version where the problem does not ever occur was 0.70.0.10.

**Comment by ashemedai** [comment]
@thebertster Oh! Very good point. Just tested again and you are right. I guess my mouse was outside the main window confines for those version. Just tested again and those indeed also fail. I'll adjust my table above.

**Comment by boba23** [comment]
This really seems to be a weird one. One thing for sure ... it's not AMD related. I have the issue with the latest Kitty version on my HP Spectre x360, Intel Ice Lake CPU, integrated IRIS Graphics, as well. 
Testing right now on my PC, i7-8700K, and what should I say ... no problems here. I can happily click Change Settings, the dialogue pops up without issues. 
Will try again on my Notebook with a clean profile, though they should be pretty much the same, since I copied them over to have the same sessions on both machines.

**Comment by n0rc** [comment]
I just tested KiTTY v0.74.0.6:
All three 32 bit release versions on FOSSHUB ([classic](https://www.fosshub.com/KiTTY.html?dwl=kitty-0.74.0.6.exe), [portable](https://www.fosshub.com/KiTTY.html?dwl=kitty_portable-0.74.0.6.exe), [no compression](https://www.fosshub.com/KiTTY.html?dwl=kitty_nocompress-0.74.0.6.exe)) crash.
However, the [latest builds on GitHub Actions](https://github.com/cyd01/KiTTY/actions/runs/235759383) work fine, both 32 bit and 64 bit.

**Comment by kissglay** [comment]
> I just tested KiTTY v0.74.0.6:
> All three 32 bit release versions on FOSSHUB ([classic](https://www.fosshub.com/KiTTY.html?dwl=kitty-0.74.0.6.exe), [portable](https://www.fosshub.com/KiTTY.html?dwl=kitty_portable-0.74.0.6.exe), [no compression](https://www.fosshub.com/KiTTY.html?dwl=kitty_nocompress-0.74.0.6.exe)) crash.
> However, the [latest builds on GitHub Actions](https://github.com/cyd01/KiTTY/actions/runs/235759383) work fine, both 32 bit and 64 bit.

Seems like it's true. Even the one from [https://github.com/cyd01/KiTTY/releases/tag/v0.74.0.6](url) crashes, but the one from Action is good.

**Comment by mickeyreg** [comment]
Hi,

Tested: 
https://github.com/cyd01/KiTTY/actions/runs/251655889 (0.74.0.7)
https://github.com/cyd01/KiTTY/actions/runs/251656811 (0.74.0.7)

and the one mentioned above:
https://github.com/cyd01/KiTTY/actions/runs/235759383 (0.74.0.6)

All three works properly.

But the last version downloaded from FossHub (today: 0.74.0.7) CRASHES.

So, it seems to be connected with the compiler used od compiler settings?

Regards,
Mickey

**Comment by cyd01** [comment]
Someone with the bug, can test this [build](https://www.9bis.net/kitty/files/kitty-debug.exe) ?

**Comment by mickeyreg** [comment]
> Someone with the bug, can test this [build](https://www.9bis.net/kitty/files/kitty-debug.exe) ?

Crashes :(

**Comment by nanaya** [comment]
Interestingly I'm not getting crashes at all on Windows 8.1.

- cpu: AMD Ryzen 5 3600
- os: Windows 8.1 (build 9600)
- kitty versions: 0.73.2.12 (portable fosshub), 0.74.1.1 (portable fosshub), [kitty-debug above](#https://github.com/cyd01/KiTTY/issues/113#issuecomment-693651851), 0.74.0.6 (github actions x64)

**Comment by mickeyreg** [comment]
Hi cyd,

I've downloaded kitty-debug.exe again and it again crashed on my PC.

Tested versions, both crashes:

![image](https://user-images.githubusercontent.com/5645076/93583846-bf709280-f9a4-11ea-8fa8-dd0b980f4ed9.png)

![image](https://user-images.githubusercontent.com/5645076/93583721-8f28f400-f9a4-11ea-95e2-701d832ac09f.png)

Regards,
Mickey

**Comment by cyd01** [comment]
Let's have some more tests.
1. Get that [build](https://www.9bis.net/kitty/files/kitty-debug.exe).
1. Add these lines
```
[KiTTY]
debug=yes
```
to your **%APPDATA%/KiTTY/kitty.ini** configuration file.

3. Run KiTTY up to crash. A kitty.log file is created near you kitty.exe binary.
4. Send me the last lines of that file

**Comment by cyd01** [comment]
Strange.
It points a PuTTY code line, not a KiTTY one. I've remove it (it is about noise, it seems not necessary). Can you test the new build again ?

**Comment by cyd01** [comment]
Stange.
It points a PuTTY code line, not a KiTTY one.
I've removed it. It is about noise and seems not necessary.
Please test this new [build](https://www.9bis.net/kitty/files/kitty-debug.exe).

**Comment by mickeyreg** [comment]
Hi Cyd,

![image](https://user-images.githubusercontent.com/5645076/93707307-6bdd8080-fb2d-11ea-951d-72a0ef37bf0b.png)
The strangest version I've tested: On one connection works fine (FreeBSD), on the other crashes (OpenWRT).

The version which I'm using now:
![image](https://user-images.githubusercontent.com/5645076/93707324-7ef05080-fb2d-11ea-9664-1e5a9f7def35.png)
This one works fine. No crash since 4 days.

Regards,
Mickey

**Comment by cyd01** [comment]
Of course it is possible. But I really need to find where the software crashes to fix it in all cases.
Can you send me the kitty.log created with very last [debug build](https://www.9bis.net/kitty/files/kitty-debug.exe) ?

**Comment by mickeyreg** [comment]
Full log from the session when the app crashed after third or fourth click on "Change Settings...": [https://misiak.mini.net.pl/~marcinkk/kitty/kitty.log.zip](url)

**Comment by cyd01** [comment]
Unlike it was suggested previously in the discussion, crash has nothing to do with **IDM_RECONF** and **WM_MOUSEMOVE** Windows event.
Your log indicate clean return after these event management. So that I've added more logging to detect the real event id that cause the crash.
Here is a new [debug build](https://www.9bis.net/kitty/files/kitty-debug.exe).

**Comment by mickeyreg** [comment]
A little problem... This build does not want to crash :(

**Comment by cyd01** [comment]
Interesting.
I think this [one](https://www.9bis.net/kitty/files/kitty-debug.exe) should crash. 
I've located the problem. It is caused by cryptographic functions. Not in KiTTY source code, but in PuTTY original code: maybe in certain circumstances (hardware+OS+compiler) there is an issue with some system call. I'm trying to find the exact line that makes the crash. Thank for your help.

**Comment by mickeyreg** [comment]
Yes. Beautifully crashes on first try :)

https://misiak.mini.net.pl/~marcinkk/kitty/kitty.log.20200922.zip

**Comment by cyd01** [comment]
Many thanks.
I've found the exact line that make the crash. It is located in file `sshprng.c`, in function `void prng_add_entropy(prng *pr, unsigned source_id, ptrlen data) ;`
at line 
```c
put_datapl(pi->collectors[index], data);
```
It is a part of the code I've never look at. So I don't exactly know what it is for (except that add_entropy certainly means adding some salt to cryptographic protocol).
Before doing anything else, I need to known the content of the variables that appear in the code.

Can you help me again by running this [build](https://www.9bis.net/kitty/files/kitty-debug.exe).
I hope it will be the last one.
Thanks.

**Comment by mickeyreg** [comment]
No problem :) Just don't be angry if I reply a little late :)

https://misiak.mini.net.pl/~marcinkk/kitty/kitty.log.20200923.zip

Like in previus pack, 3 logs inside. The log in file kitty.log is from the session, when the first try to "Change Settings..." did not crash.

**Comment by cyd01** [comment]
Confirmation! 
Thanks for your help.
To be honest I don't really know how to fix the issue definitely. I can propose a workaround by activating a feature that I've found in PuTTY source code.

Get this [last build](https://www.9bis.net/kitty/files/kitty-debug.exe).
Normally it will crash ...

Now add these lines to your `%APPDATA/KiTTY/kitty.ini` configuration file:
```
[Debug]
randomactive=no
```
With this setting ... it should not crash anymore.
Hope it will help.

**Comment by mickeyreg** [comment]
It works exactly as you described. Normally crashes, but these lines in the configuration file make the application stable :)

**Comment by n0rc** [comment]
> Confirmation!
> Thanks for your help.

Thanks a lot for working an this!

> To be honest I don't really know how to fix the issue definitely. I can propose a workaround by activating a feature that I've found in PuTTY source code.

What exactly does this feature do?

I'm asking because I don't understand why it would be necessary to change PuTTY source code at all: The earlier builds from GitHub Actions did work without any PuTTY source code changes and PuTTY itself also does not crash.

Does PuTTY's `prng_add_entropy` somehow affect code from KiTTY?

**Comment by devilcrayon** [comment]
Also confirming that with the debug section added to `kitty.ini` the debug build no longer crashes for me.

**Comment by cyd01** [comment]
I'm preparing a definitive fix (i.e. without need to add the Debug section).
Can you do something for me ?
Get that [build](https://www.9bis.net/kitty/files/kitty-debug.exe) and make a [memory dump](https://www.9bis.net/kitty/#!pages/Savedump.md) on the system that usually crash of course.
It will produce a `kitty.dmp` file. Please attach it to your response.
Thanks.

**Comment by mickeyreg** [comment]
Nothing happens when I press "Shift+F8" on my system :(

**Comment by cyd01** [comment]
Try to put 

    [KiTTY]
    debug=yes
only, in you config file, nothing else.

**Comment by mickeyreg** [comment]
Works :)

https://misiak.mini.net.pl/~marcinkk/kitty/kitty.dmp.20200925.zip

**Comment by devilcrayon** [comment]
Here's another:
[kitty.dmp.zip](https://github.com/cyd01/KiTTY/files/5284001/kitty.dmp.zip)

**Comment by cyd01** [comment]
All of these exports appears to came from a Intel X86 system !
So that I can't use the system detection to fix the issue. I will activate the PuTTY switch for everyone.

**Comment by cyd01** [comment]
> I'm asking because I don't understand why it would be necessary to change PuTTY source code at all: The earlier builds from GitHub Actions did work without any PuTTY source code changes and PuTTY itself also does not crash.

I said the problem seems come from a bad combination of several things
- system (AMD)
- OS (Win 10)
- Compiler

In Github Actions, the compilation is done on a Linux box with gcc 9.3 into a Docker container.

But I work on a PC, without Docker.   
So that my building and publish chain for official releases is completely different.   
It is based on Windows 10 with gcc 6.3. It explains the differences.

**Comment by ssbssa** [comment]
Is there a reason why you don't use a more recent gcc version?

**Comment by n0rc** [comment]
> I said the problem seems come from a bad combination of several things (…)

Yes, I understand that the build chain environment causes the problem.

I'm just curious what exactly is this feature you mentioned and if changes in the [mentioned part](https://github.com/cyd01/KiTTY/issues/113#issuecomment-697162848) of PuTTY's source code were necessary for the workaround.

**Comment by cyd01** [comment]
I'm not sure it will. 
I've just upgraded to gcc 9.2.

**Comment by kissglay** [comment]
It seems fixed in 0.74.2.3. Thx.

**Comment by n0rc** [comment]
@cyd01 I looked at your commit now and saw the `RandomActiveFlag` you introduced to fix the bug.
You seem to use it to [effectively ignore](https://github.com/cyd01/KiTTY/blob/c4b6dabb2814403fb2f4ddc549d06a48bc953c10/0.74_My_PuTTY/sshrand.c#L45) PuTTY code that [adds entropy to the PRNG](https://github.com/cyd01/KiTTY/blob/c4b6dabb2814403fb2f4ddc549d06a48bc953c10/0.74_My_PuTTY/sshrand.c#L50) by setting `RandomActiveFlag` to 0 [per default](https://github.com/cyd01/KiTTY/blob/c4b6dabb2814403fb2f4ddc549d06a48bc953c10/kitty_commun.c#L93).

You [mentioned earlier](https://github.com/cyd01/KiTTY/issues/113#issuecomment-695206800) that you think this code is not necessary. Can you please explain why you think so?

Also, as I meantioned before, it seems that this bug was *not* caused by PuTTY's code but by the build chain used: PuTTY itself did never crash with the code and the *same KiTTY code without your latest modifications* ran without issues when using the GitHub Actions builds.

Why do you think changing PuTTY's code here (especially by ignoring code related to cryptographic functionality) is the right approach to fix this bug rather than fixing the build chain issues?
